CASSANDRA-18646 Azure snitch

smiklosovic opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

jacek-lewandowski commented on src/java/org/apache/cassandra/locator/AzureSnitch.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    public String getRack(InetAddressAndPort endpoint)
```
This method, and `getDatacenter` method, are identical in all snitch implementations. The only difference is that it returns a value of a field `rack` and, say in EC2, it returns value of a field `ec2zone`. Could you extract simple method like `getRackInternal() { return rack }`, define it abstract in `AbstractNetworkTopologySnitch` and move the implementation of `getRack(endpoint)` to `AbstractNetworkTopologySnitch`? You can do the same for `getDatacenter(endpoint)` method, and do that for all the snitch implementations.

smiklosovic committed:
Add AzureSnitch

As we were implementing the snitch itself, we noticed that the constructors
for cloud-based snitches are unnecessarilly complicated and
we took the opportunity to make them simpler.

patch by Stefan Miklosovic; reviewed by German Eichberger and Jacek Lewandowski for CASSANDRA-18646

xgerman commented on src/java/org/apache/cassandra/locator/AzureSnitch.java:
```diff
@@ -0,0 +1,109 @@
+        JsonNode jsonNode = JsonUtils.JSON_OBJECT_MAPPER.readTree(response);
+
+        JsonNode location = jsonNode.get("location");
+        JsonNode zone = jsonNode.get("zone");
```
Azure being azure there is another way to distribute nodes called fault domains (see https://learn.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-manage-fault-domains) so if zone is empty you need to check `platformFaultDomain` https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux

xgerman commented on src/java/org/apache/cassandra/locator/AzureSnitch.java:
```diff
@@ -0,0 +1,109 @@
+        if (location == null || location.isNull() || location.asText().isEmpty())
+            datacenter = DEFAULT_DC;
+        else
+            datacenter = location.asText();
```
not sure if having multiple data centers in the same location is a goal. If so, this would not work for that scenario.

smiklosovic replied:
@xgerman I am not sure what you mean by this. Can you elaborate?

xgerman replied:
Assume you have two data centers DC1 and DC2 both in the same region, say useast2. Then `datacenter` would be for both useast2...

smiklosovic replied:
resolved privately, in practice one is supposed to use dc_suffix for this kind of scenario

xgerman commented on conf/cassandra.yaml:
```diff
@@ -1302,6 +1302,10 @@ slow_query_log_timeout: 500ms
 #
+# AzureSnitch:
+#    Gets datacenter from 'location' and rack from 'zone' fields of 'compute' object
+#    from instance metadata service.
```
If availability zone is not enabled it will use fault domains and get the respective value from

jacek-lewandowski commented on src/java/org/apache/cassandra/locator/AzureSnitch.java:
```diff
@@ -0,0 +1,128 @@
+        if (location == null || location.isNull() || location.asText().isEmpty())
+            datacenter = DEFAULT_DC;
+        else
+        {
```
nit: braces seems redundant

smiklosovic replied:
done

jacek-lewandowski commented on conf/cassandra.yaml:
```diff
@@ -1302,6 +1302,11 @@ slow_query_log_timeout: 500ms
 #    Snitch for getting dc and rack of a node from metadata service of Alibaba cloud.
 #    This snitch that assumes an ECS region is a DC and an ECS availability_zone is a rack.
 #
+# AzureSnitch:
```
nit: 
```
# AzureSnitch:
#    Gets datacenter from 'location' and rack from 'zone' fields of 'compute' object
#    from instance metadata service. If the availability zone is not enabled, it will use the fault
#    domain and get its respective value.
#
```

jacek-lewandowski reviewed (approved):
+1, please make sure CI is passing before merge

smiklosovic committed:
DO NOT COMMIT
